module/zfs/dbuf.c

```diff
@@ -1218,17 +1235,105 @@ dbuf_read_done(zio_t *zio, const zbookmark_phys_t *zb, const blkptr_t *bp,
 		arc_buf_freeze(buf);
 		db->db_freed_in_flight = FALSE;
 		dbuf_set_data(db, buf);
-		db->db_state = DB_CACHED;
+		DBUF_STATE_CHANGE(db, =, DB_CACHED, "freed in flight");
 	} else {
 		/* success */
 		ASSERT(zio == NULL || zio->io_error == 0);
 		dbuf_set_data(db, buf);
-		db->db_state = DB_CACHED;
+		DBUF_STATE_CHANGE(db, =, DB_CACHED, "successful read");
 	}
 	cv_broadcast(&db->db_changed);
 	dbuf_rele_and_unlock(db, NULL, B_FALSE);
 }
 
+/*
+ * Shortcut for performing reads on bonus dbufs.  Returns whether any action
+ * was taken.
+ */
+static boolean_t
+dbuf_read_bonus(dmu_buf_impl_t *db, dnode_t *dn, uint32_t flags, int *err)
```

Comment:
Can you restructure this the traditional way and return the error, instead of assigning it to the passed `*err` argument.  Zero on success, otherwise a useful errno.  I'd suggest `EINVAL` for the `db->db_blkid != DMU_BONUS_BLKID` case.